aten/src/ATen/native/cuda/Col2Im.cu

```diff
@@ -0,0 +1,307 @@
+  int64_t kernel_height = kernel_size[0];
+  int64_t kernel_width = kernel_size[1];
+  int64_t dilation_height = dilation[0];
+  int64_t dilation_width = dilation[1];
+  int64_t pad_height = padding[0];
+  int64_t pad_width = padding[1];
+  int64_t stride_height = stride[0];
+  int64_t stride_width = stride[1];
+
+  col2im_shape_check(
+      input_,
+      Tensor(),
+      output_height,
+      output_width,
+      kernel_height,
+      kernel_width,
+      dilation_height,
+      dilation_width,
+      pad_height,
+      pad_width,
+      stride_height,
+      stride_width);
+
+  Tensor input = input_.contiguous();
+
+  bool batched_input = true;
+  if (input.dim() == 2) {
+    // Force batch
+    batched_input = false;
+    input.resize_({1, input.size(0), input.size(1)});
+  }
+
+  int64_t batch_size = input.size(0);
+  int64_t n_input_plane = input.size(1);
+  int64_t n_output_plane = n_input_plane / (kernel_width * kernel_height);
+
+  output.resize_({batch_size, n_output_plane, output_height, output_width});
+  output.zero_();
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "col2im_out_cuda", [&] {
+    using accscalar_t = at::acc_type<scalar_t, true>;
+
+    Tensor input_n;
+    Tensor output_n;
+
+    int64_t height_col = (output_height + 2 * pad_height -
+                          (dilation_height * (kernel_height - 1) + 1)) /
+            stride_height +
+        1;
+    int64_t width_col = (output_width + 2 * pad_width -
+                         (dilation_width * (kernel_width - 1) + 1)) /
+            stride_width +
+        1;
+
+    for (int64_t elt = 0; elt < batch_size; elt++) {
+      input_n = input.select(0, elt);
+      output_n = output.select(0, elt);
+
+      col2im<scalar_t, accscalar_t>(
+          at::cuda::getCurrentCUDAStream(),
+          input_n.data<scalar_t>(),
+          n_output_plane,
+          output_height,
+          output_width,
+          height_col,
+          width_col,
+          kernel_height,
+          kernel_width,
+          pad_height,
+          pad_width,
+          stride_height,
+          stride_width,
+          dilation_height,
+          dilation_width,
+          output_n.data<scalar_t>());
+    }
+
+    if (!batched_input) {
+      output.resize_({n_output_plane, output_height, output_width});
+    }
+  });
+}
+
+void col2im_backward_out_cuda_template(
+    Tensor& grad_input,
+    Tensor& grad_output,
+    IntArrayRef kernel_size,
+    IntArrayRef dilation,
+    IntArrayRef padding,
+    IntArrayRef stride) {
+  // im2col_out_cuda checks size of kernel_size, dilation, padding and stride
+  grad_input = im2col_cuda(
+      grad_output, kernel_size, dilation, padding, stride);
+}
+
+} // namespace
+
+Tensor& col2im_out_cuda(
+    Tensor& output,
+    const Tensor& input_,
+    IntArrayRef output_size,
+    IntArrayRef kernel_size,
+    IntArrayRef dilation,
+    IntArrayRef padding,
+    IntArrayRef stride) {
+  Tensor input = input_;
```

Comment:
That's weird. Why is `col2im_out_cuda_template` taking a mutable reference for input? We shouldn't be mutating it right?

Reply:
you're right ... I don't remember why I did it .. but I will change that to const here. thanks !

Reply:
mmm actually it seems I created the `out` function just for cuda for `col2im` .. but actually I am not using that in the code ... I will try to create `out` functions for col2im and im2col